lib/ecto/changeset.ex

```diff
@@ -1349,6 +1349,67 @@ defmodule Ecto.Changeset do
     end
   end
 
+  @doc """
+  Validates (unreliably) that conflicting records do not exist in the database.
+
+  This is not reliable because conflicting records could be inserted by another
+  process after this validation runs. Use a constraint for a reliable check.
+
+  However, in most cases when there are conflicts, they will already exist by
+  the time of validation. Checking for them in the validation phase may allow
+  users to fix errors like "username is already taken" at the same time as
+  errors like "email can't be blank", instead of fixing each constraint error
+  the database raises, one at a time, after the validations pass.
+
+  To use this, pass a function that finds existing records with which this changeset
+  would conflict, according to a unique or exclusion constraint. If the function
+  returns any an empty enumerable, the validation passes, otherwise it fails.
+
+  The `field_name` is where the error message (if any) will be placed.
+
+  ## Options
+
+    * `:message` - the message on failure, defaults to "has already been taken"
+
+  ## Examples
+
+      validate_no_conflicts_unreliably(changeset, :title, &find_posts_with_title/1)
+      validate_no_conflicts_unreliably(changeset, :title, &find_posts_with_title/1, message: "not original")
+      validate_no_conflicts_unreliably(changeset, :base, &find_overlapping_rentals/1, message: "an existing rental overlaps these dates")
+
+  A finder function might look like:
+
+  defp find_posts_with_title(changeset) do
+    title = get_field(changeset, :title)
+    if is_nil(title) do
+      # no title, so there can be no conflicting records
+      []
+    else
+      MyApp.Repo.all(from p in Post, where: e.title == ^title)
+    end
+  end
+  """
+  @spec validate_no_conflicts_unreliably(t, atom, (t -> list), String.t) :: t
+  def validate_no_conflicts_unreliably(changeset, field_name, conflict_finder_func, opts \\ []) do
+    conflicts = changeset
+    |> conflict_finder_func.()
+    |> Enum.reject(fn (conflict) ->
+      conflict.id == get_field(changeset, :id)
```

Comment:
Is it a problem to assume there's a unique `:id` field? If so, should this "a record doesn't conflict with itself" check also be handled by the `conflict_finder_func`?